pyevmasm/evmasm.py

```diff
@@ -1090,6 +596,288 @@ def assemble_hex(asmcode, pc=0, fork=DEFAULT_FORK):
         return '0x' + hexlify(b''.join([x.bytes for x in asmcode])).decode('ascii')
     return '0x' + hexlify(assemble(asmcode, pc=pc, fork=fork)).decode('ascii')
 
+
+class InstructionTable():
+    """
+    EVM Instruction factory
+        Implements an immutable, iterable instruction LUT that can be indexed by both mnemonic or opcode.
+
+    Example::
+
+        >>> from pyevmasm import instruction_tables
+        >>> instruction_table = instruction_tables['byzantium']
+        >>> instruction_table[0]
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> instruction_table['STOP']
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> i = instruction_table.__iter__()
+        >>> i.__next__()
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> i.__next__()
+        Instruction(0x1, 'ADD', 0, 2, 1, 3, 'Addition operation.', None, 0)
+        >>> i.__next__()
+        Instruction(0x2, 'MUL', 0, 2, 1, 5, 'Multiplication operation.', None, 0)
+        >>> i.__next__()
+        Instruction(0x3, 'SUB', 0, 2, 1, 3, 'Subtraction operation.', None, 0)
+
+    """
+    __slots__ = ('_previous_fork', '_ilist', '_cache')
+    def __init__(self, *args, **kwargs):
+        self._previous_fork = kwargs.get('previous_fork', None)
+        self._ilist = {}
+        self._ilist.update(args[0])
+        self._cache = {}
+
+    def _search_by_opcode(self, k):
+        try:
+            return (k, *self._ilist[k])
+        except KeyError:
+            if self._previous_fork is not None:
+                return self._previous_fork._search_by_opcode(k)
+        raise KeyError(k)
+
+    def _search_by_name(self, k):
+        for opcode, (name, operand_size, pushes, pops, gas, description) in self._ilist.items():
```

Comment:
uhm... wouldn't it be better to have two dictionaries? one `opcode -> all_instruction_data` and second `name -> opcode`?

As it seems we may search each time an instruction is going to be fetched and so iterate through all the big instruction dict each time?

Reply:
I changed it so each version/fork of the instruction list has an opcode - > name map built eagerly at __init__. We'll see..